python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -0,0 +1,1023 @@
+    implementation of hashing and equality schemes, child classes need
+    to provide a certain amount of metadata when they are defined.
+    Specifically, the ``_non_child`` attribute must list, in-order,
+    the names of the slots that are passed to the constructor. The
+    constructor must take arguments in the order ``(*_non_child,
+    *children).``
+    """
+
+    __slots__ = ("dtype", "hash_value", "repr_value")
+    dtype: plc.DataType
+    """Data type of the expression."""
+    hash_value: int
+    """Caching slot for the hash of the expression."""
+    repr_value: str
+    """Caching slot for repr of the expression."""
+    children: tuple[Expr, ...] = ()
+    """Children of the expression."""
+    _non_child: ClassVar[tuple[str, ...]] = ("dtype",)
+    """Names of non-child data (not Exprs) for reconstruction."""
+
+    # Constructor must take arguments in order (*_non_child, *children)
+    def __init__(self, dtype: plc.DataType) -> None:
+        self.dtype = dtype
+
+    def _ctor_arguments(self, children: Sequence[Expr]) -> Sequence:
+        return (*(getattr(self, attr) for attr in self._non_child), *children)
+
+    def get_hash(self) -> int:
+        """
+        Return the hash of this expr.
+
+        Override this in subclasses, rather than __hash__.
+
+        Returns
+        -------
+        The integer hash value.
+        """
+        return hash((type(self), self._ctor_arguments(self.children)))
+
+    def __hash__(self):
+        """Hash of an expression with caching."""
+        try:
+            return self.hash_value
+        except AttributeError:
+            self.hash_value = self.get_hash()
+            return self.hash_value
+
+    def is_equal(self, other: Any) -> bool:
+        """
+        Equality of two expressions.
+
+        Override this in subclasses, rather than __eq__.
+
+        Parameter
+        ---------
+        other
+            object to compare to
+
+        Returns
+        -------
+        True if the two expressions are equal, false otherwise.
+        """
+        if type(self) is not type(other):
+            return False
+        return self._ctor_arguments(self.children) == other._ctor_arguments(
+            other.children
+        )
+
+    def __eq__(self, other):
+        """Equality of expressions."""
+        if type(self) != type(other) or hash(self) != hash(other):
+            return False
+        else:
+            return self.is_equal(other)
+
+    def __ne__(self, other):
+        """Inequality of expressions."""
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        """String representation of an expression with caching."""
+        try:
+            return self.repr_value
+        except AttributeError:
+            args = ", ".join(f"{arg!r}" for arg in self._ctor_arguments(self.children))
+            self.repr_value = f"{type(self).__name__}({args})"
+            return self.repr_value
+
+    # TODO: return type is a lie for Literal
```

Comment:
Should we mark the return as a union, or is it possible that the list could extend further and this really is polymorphic amongst children?